# Scripts #



1. **cal_fastxLenStats.pl** : calculate total length, no of records, min and max lengths in a fasta file.
2. **fastq_QF.pl** : Takes in two fastq files (paired), compare their Quality and filter if mean Quality is <20 for each record, and filters them. 
3. **fasta_appendchr.pl** : Appends "chr" after">" in fasta file \(BioPerl\).
4. **tif2jpeg.sh** : Converts ".tif" images into ".jpeg" using sips image conversion tool in MacOS.
5. **compareLines.pl** : Compares two given files line by line and print common and unique ones, Output includes three coloumns, 1st Col (uniq in A), 2nd (Uniq in B), 3rd (Common in Aand B).
6. **hmmscan_domtblout2tab.pl** : parse hmmscan domtbl outout to convert it to tab seperated file (easy for downstream analysis).
7. **fasta_filterBylen.pl** : filter multi fasta file based on their lenghts.
8. **extract_rec_fastx.py** : extract random records from (fasta / fastq ) file  [PYTHON SCRIPT].
9. **gtf2fasta.pl** : convert gtf to fasta format.
10. **split_fileByLines.pl** : split a file based on no of lines in each splitted file.
11. **fastqlen.pl** : Ouput the record id and len for each record in multi fastq file.
